VisualizationBase/src/items/ViewItem.cpp

```diff
@@ -102,38 +107,58 @@ const QPoint ViewItem::positionOfNode(Model::Node *node) const
 	return QPoint(-1, -1);
 }
 
-QVector<QVector<Model::Node*>> ViewItem::nodesGetter()
+const QPoint ViewItem::positionOfItem(Item *item) const
 {
-	return nodes_;
+	auto vref = DCast<VViewItemNode>(item);
+	if (vref) return positionOfNode(vref->node());
+	else return QPoint(-1, -1);
 }
 
-void ViewItem::ensureColumnExists(int column)
+Model::Node* ViewItem::nodeAt(int column, int row)
 {
-	if (nodes_.size() <= column)
-		nodes_.resize(column + 1);
+	if (column < 0 || column >= nodes_.size())
+		return nullptr;
+	if (row < 0 || row >= nodes_[column].size())
+		return nullptr;
+	return nodes_[column][row];
 }
 
-QString ViewItem::debugNodes()
+void ViewItem::addSpacing(int column, int row, Model::Node* spacingTarget)
 {
-	//Prints the node grid as a debug string
-	QString result;
-	int maxHeight = 0;
-	for (int i = 0; i < nodes_.size(); i++)
-		if (nodes_.at(i).size() > maxHeight)
-			maxHeight = nodes_.at(i).size();
+	insertViewItemNode(ViewItemNode::withSpacingTarget(spacingTarget), column, row);
+}
 
-	for (int row = 0; row < maxHeight; row++)
+void ViewItem::updateGeometry(int availableWidth, int availableHeight)
+{
+	for (auto node : allNodes())
 	{
-		for (int col = 0; col < nodes_.size(); col++)
-		{
-			if (nodes_.at(col).size() > row)
-				result += nodes_.at(col).at(row) ? (nodes_.at(col).at(row)->typeName() + " ") : "null ";
-			else
-				result += "null ";
-		}
-		result += "\n";
+		auto asViewItemNode = DCast<ViewItemNode>(node);
+		Q_ASSERT(asViewItemNode);
+		auto item = DCast<VViewItemNode>(findVisualizationOf(asViewItemNode));
+		Q_ASSERT(item);
+		if (item->currentFormIndex() == 1) item->determineSpacing();
 	}
-	return result;
+	Super::updateGeometry(availableWidth, availableHeight);
+}
+
+void ViewItem::insertViewItemNode(ViewItemNode *node, int column, int row)
+{
+	//First, make sure the current grid is big enough to fit the node
+	ensureColumnExists(column);
+	if (nodes_[column].size() <= row)
+		nodes_[column].resize(row + 1);
```

Comment:
Shouldn't you somehow merge this in the `ensureColumnExists` method and rename the method to `ensurePositionExists` or something?